opencti-platform/opencti-graphql/src/schema/schema-attributes.ts

```diff
@@ -26,94 +27,99 @@ export const schemaAttributesDefinition = {
-    const currentMap = this.types[type];
-    if (currentMap) {
-      values.forEach((v) => currentMap.set(v));
-    } else {
-      this.types[type] = new Map(values.map((c) => [c, undefined]));
-    }
-  },
-  get(type: string): string[] {
-    return Array.from(this.types[type].keys());
-  },
+  upsertByEntity: new Map<string, string[]>(),
 
   // attributes
   registerAttributes(entityType: string, attributes: AttributeDefinition[]) {
-    // Check the homogeneity of attribute types
-    const allAttributes = this.getAllAttributes();
+    const directAttributes = this.attributes[entityType] ?? new Map<string, AttributeDefinition>();
+
+    // Register given attribute
+    const allAttributes = Object.values(this.attributes);
     attributes.forEach((attribute) => {
-      const existingAttribute = allAttributes.find((a) => a.name === attribute.name);
+      // Check the homogeneity of attribute types
+      const existingAttribute = allAttributes.find((a) => a.get(attribute.name))?.get(attribute.name); // Maybe better way ?
       if (existingAttribute && existingAttribute.type !== attribute.type) {
-        throw UnsupportedError('You can\'t have two attribute with the same name and a different type in the platform', {
+        throw UnsupportedError('You can\'t have two attributes with the same name and a different type in the platform', {
           existingAttribute,
           attribute,
         });
       }
-      this.attributes_by_name[attribute.name] = attribute;
-      // Generate map by types
-      this.attributes_by_types[attribute.type].set(attribute.name);
-      // Generate map by multiple
-      if (attribute.multiple) {
-        this.attributes_multiple.set(attribute.name);
-      }
-      // Generate map of upsert by entity type
-      if (attribute.upsert) {
-        const currentList = this.upsert_by_entity.get(entityType);
-        if (currentList) {
-          currentList.push(attribute.name);
-        } else {
-          this.upsert_by_entity.set(entityType, [attribute.name]);
-        }
+      // Check duplicate attributes
+      if (directAttributes.has(attribute.name)) {
+        throw UnsupportedError('You can\'t register two attributes with the same name on an entity', {
+          attributeName: attribute.name,
+          entityType
+        });
       }
+
+      directAttributes.set(attribute.name, attribute);
     });
-    this.attributes[entityType] = attributes;
+    this.attributes[entityType] = directAttributes;
+
+    this.computeCache(entityType);
   },
 
-  getAttribute(name: string): AttributeDefinition | null {
-    return this.attributes_by_name[name];
+  // Extract this method to be call in all methods
+  // When an entity not register any relations, the relations for this entity is not computed
+  // Call only in register mechanism when all the entities will be migrated
+  computeCache(entityType: string) {
+    if (this.attributesCache.has(entityType)) return;
```

Comment:
Thats a uncommon way to do a caching function.
Should just be getAttributesFromCache() 
If Not => compute and result
If Yes, result

Reply:
Add registration for each existing type to remove cache